File: 144.Binary Tree Preorder Traversal/c++/main.cpp
/**144. Binary Tree Preorder Traversal
 * Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree [1,null,2,3],
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?*/
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution//非递归前续遍历
{
public:
    vector<int> preorderTraversal(TreeNode* root)
    {
        vector<int> ret;
        if(root == NULL)
            return ret;
        //声明一个栈
        stack<TreeNode*> s;//辅助

        s.push(root);//根节点入栈

        while(!s.empty())
        {
            TreeNode* p = s.top();
            s.pop();
            ret.push_back(p->val);
            if(p->right != NULL)
                s.push(p->right);//先压入右孩子，这样每次取出的都是左孩子(如果存在的话)。
            if(p->left != NULL)
                s.push(p->left);
        }

        return ret;
    }
};

int main() {
    std::cout << "Hello, World!" << std::endl;
    return 0;
}

File: 144.Binary Tree Preorder Traversal/main.java
// 144. Binary Tree Preorder Traversal
// Given a binary tree, return the preorder traversal of its nodes' values. 
//
// Example: 
//
// 
//Input: [1,null,2,3]
//   1
//    \
//     2
//    /
//   3
//
//Output: [1,2,3]
// 
//
// Follow up: Recursive solution is trivial, could you do it iteratively? 
// Related Topics Stack Tree


//leetcode submit region begin(Prohibit modification and deletion)

import javax.swing.tree.TreeNode;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
    	//储存遍历结果
        List<Integer> ret = new ArrayList<>();

        //判空
        if(root == null){
            return ret;
        }

        //辅助栈
        Stack<TreeNode> helper = new Stack<TreeNode>();
        //入栈根节点
        helper.push(root);

        //循环直到栈为空
        while(!helper.isEmpty()){
        	//取栈顶指针
            TreeNode current = helper.pop();
            //把当前值加入遍历结果
            ret.add(current.val);
            /**
            前序遍历顺序为根左右,因为栈是先进后出,所以要先访问的节点最后入*/
            //先入栈右节点
            if(current.right != null){
                helper.push(current.right);
            }
            //后入栈左节点
            if(current.left != null){
                helper.push(current.left);
            }
        }
        return ret;
    }
}
//leetcode submit region end(Prohibit modification and deletion)